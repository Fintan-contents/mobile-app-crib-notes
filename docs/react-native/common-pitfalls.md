---
title: ハマりがちな落とし穴
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
