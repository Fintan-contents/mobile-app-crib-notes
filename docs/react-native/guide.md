---
title: 何らかのガイドを書きたい気持ちはある
---
