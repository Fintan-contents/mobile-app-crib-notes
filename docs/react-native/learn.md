---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。

ReactでWebアプリケーションを実装できるがReact Nativeでモバイルアプリケーションを実装したことがないアプリケーション開発者を対象として、次のことについて学ぶことを目的としています。

- ReactとReact Nativeの違い
- React Nativeで作成したモバイルアプリケーションの動作確認方法

また、作成したモバイルアプリケーションの配布方法やテスト手法についてはこのコンテンツでは触れません。

<!-- textlint-disable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Getting Started',
    to: '/react-native/learn/getting-started',
    summary: <>React Nativeでのアプリ開発に必要な環境を用意します。</>
  },
  {
    title: 'アプリの実装前に',
    to: '/react-native/learn/basic-concepts',
    summary: <>React Nativeでアプリを実装し始める前に最低限必要となる前提知識や、React NativeとReact Navigationの基本的な知識について学びます。</>
  },
  {
    title: 'ToDoアプリの実装',
    to: '/react-native/learn/todo-app',
    summary: <>React Nativeで簡単なToDoアプリを実装します。</>
  },
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->
