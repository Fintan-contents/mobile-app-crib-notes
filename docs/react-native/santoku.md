---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

React Nativeサンプルアプリケーションの設計ドキュメントです。

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
