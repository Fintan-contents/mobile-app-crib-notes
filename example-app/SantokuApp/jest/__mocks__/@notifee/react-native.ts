import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

// テストケースごとにモックは初期化しておく。
beforeEach(() =>
  Object.values(mock).forEach(fn => {
    if (jest.isMockFunction(fn)) {
      fn.mockClear();
    }
  }),
);

export default mock;
