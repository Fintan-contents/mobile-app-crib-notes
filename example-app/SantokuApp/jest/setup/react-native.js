// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
