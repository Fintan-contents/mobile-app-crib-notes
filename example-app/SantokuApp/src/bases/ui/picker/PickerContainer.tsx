import React from 'react';
import {StyleSheet, View, ViewProps} from 'react-native';
import Reanimated, {WithTimingConfig} from 'react-native-reanimated';

import {usePickerContainer} from './usePickerContainer';

export const DEFAULT_SLIDE_IN_DURATION = 300;
export const DEFAULT_SLIDE_OUT_DURATION = 300;

// React Native ReanimatedのLayout Animationsを使用すると↓の不具合が発生するため、指定できないようにしています。
// https://github.com/software-mansion/react-native-reanimated/issues/2906
export type PickerContainerProps = Omit<Reanimated.AnimateProps<ViewProps>, 'exiting' | 'entering'> & {
  isVisible: boolean;
  afterSlideIn?: (finished?: boolean) => unknown;
  afterSlideOut?: (finished?: boolean) => unknown;
  slideInDuration?: number;
  slideOutDuration?: number;
  slideInConfig?: WithTimingConfig;
  slideOutConfig?: WithTimingConfig;
};

export const PickerContainer: React.FC<React.PropsWithChildren<PickerContainerProps>> = ({
  isVisible,
  style,
  afterSlideIn,
  afterSlideOut,
  slideInDuration = DEFAULT_SLIDE_IN_DURATION,
  slideOutDuration = DEFAULT_SLIDE_OUT_DURATION,
  slideInConfig,
  slideOutConfig,
  children,
  ...animatedViewProps
}) => {
  const {isPickerVisible, transform, updateContentHeight} = usePickerContainer({
    isVisible,
    afterSlideIn,
    afterSlideOut,
    slideInDuration,
    slideOutDuration,
    slideInConfig,
    slideOutConfig,
  });
  return !isPickerVisible ? null : (
    <View style={[styles.container]} pointerEvents="box-none">
      <Reanimated.View style={[style, transform]} onLayout={updateContentHeight} {...animatedViewProps}>
        {children}
      </Reanimated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
  },
});
