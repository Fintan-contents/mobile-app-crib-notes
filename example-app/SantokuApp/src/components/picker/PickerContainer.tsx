import {useWorkletCallback} from 'framework/utilities/useWorkletCallback';
import React from 'react';
import {StyleSheet, View, ViewProps} from 'react-native';
import Reanimated, {BaseAnimationBuilder, Keyframe, SlideInDown, SlideOutDown} from 'react-native-reanimated';

export type PickerContainerProps = Omit<Reanimated.AnimateProps<ViewProps>, 'entering' | 'exiting'> & {
  isVisible: boolean;
  /**
   * iOSの場合、アニメーションが終わった後に呼び出されます。
   * Androidの場合、アニメーションが始まった時に呼び出されます。
   */
  enteringCallback?: (finished: boolean) => unknown;
  exitingCallback?: (finished: boolean) => unknown;
  /**
   * enteringに指定したAnimationBuilderなどでwithCallbackを指定しても、本コンポーネント内で上書きしているため実行できません。
   * withCallbackで実行する関数は、enteringCallbackで指定してください。
   */
  entering?: BaseAnimationBuilder | Keyframe;
  /**
   * exitingに指定したAnimationBuilderなどでwithCallbackを指定しても、本コンポーネント内で上書きしているため実行できません。
   * withCallbackで実行する関数は、exitingCallbackで指定してください。
   */
  exiting?: BaseAnimationBuilder | Keyframe;
};

export const PICKER_CONTAINER_DEFAULT_SLIDE_IN_DURATION = 300;
export const PICKER_CONTAINER_DEFAULT_SLIDE_OUT_DURATION = 300;
export const PICKER_CONTAINER_DEFAULT_ENTERING = SlideInDown.duration(PICKER_CONTAINER_DEFAULT_SLIDE_IN_DURATION);
export const PICKER_CONTAINER_DEFAULT_EXITING = SlideOutDown.duration(PICKER_CONTAINER_DEFAULT_SLIDE_OUT_DURATION);

export const PickerContainer: React.FC<PickerContainerProps> = ({
  isVisible,
  entering = PICKER_CONTAINER_DEFAULT_ENTERING,
  exiting = PICKER_CONTAINER_DEFAULT_EXITING,
  enteringCallback,
  exitingCallback,
  style,
  children,
  ...animatedViewProps
}) => {
  const composedEnteringCallback = useWorkletCallback(enteringCallback);
  const composedExitingCallback = useWorkletCallback(exitingCallback);

  return (
    <>
      {isVisible && (
        <View style={styles.container} pointerEvents="box-none">
          <Reanimated.View
            style={style}
            entering={entering.withCallback(composedEnteringCallback)}
            exiting={exiting.withCallback(composedExitingCallback)}
            pointerEvents="box-none"
            {...animatedViewProps}>
            {children}
          </Reanimated.View>
        </View>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
  },
});
