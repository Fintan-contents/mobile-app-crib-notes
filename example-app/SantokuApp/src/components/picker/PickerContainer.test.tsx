import {render} from '@testing-library/react-native';
import React from 'react';
import {ViewProps} from 'react-native';
import Reanimated, {ZoomIn, ZoomOut} from 'react-native-reanimated';

import {PICKER_CONTAINER_DEFAULT_ENTERING, PICKER_CONTAINER_DEFAULT_EXITING, PickerContainer} from './PickerContainer';

// If advancing a timer changes the state of a component, the timer must be run within an act.
// However, since act is `Thenable`, ESLint will issue a warning if you do not do something like await.
// For convenience, disable the relevant rule in this file.
/* eslint-disable @typescript-eslint/no-floating-promises */

// To get react-native-reanimated v2 animations to work with timer mocks, you need to use the `modern` timer.
// By default, Jest v26 uses `legacy` timer. So if you use `withReanimatedTimer` that use the default timer, animation does not start.
// https://jestjs.io/ja/docs/timer-mocks
jest.useFakeTimers('modern');

// TODO: Jest v27にアップデートできたら、withReanimatedTimerでテストを実装できるか検証する。
//       （JestのバージョンはExpoに依存しているので、Expoでのアップデートを待っている状態）

jest.runAllTimers();

describe('PickerContainer only with required props', () => {
  it('returns null if not visible', () => {
    const sut = render(<PickerContainer isVisible={false} />);
    // PickerContainerがnullを返していることを確認したいがうまくやる方法が見当たらないので`toJSON`でnullになることを確認する。
    expect(sut.toJSON()).toBeNull();
  });

  it('renders successfully only with required props', () => {
    const sut = render(<PickerContainer isVisible testID="containerAnimated" />);
    const animatedView = sut.getByTestId('containerAnimated');
    const animatedViewProps = animatedView.props as Reanimated.AnimateProps<ViewProps>;
    // Animated.Viewのentering/exitingをテストで実行することができなかったため、entering/exitingにデフォルトアニメーションが設定されていることのみを確認する。
    expect(sut).toMatchSnapshot('AnimatedView with visible.');
    expect(animatedView).not.toBeNull();
    expect(animatedViewProps.entering).toBe(PICKER_CONTAINER_DEFAULT_ENTERING);
    expect(animatedViewProps.exiting).toBe(PICKER_CONTAINER_DEFAULT_EXITING);

    //////////////////////////////////////////////////////////////////////////////////
    // 非表示にする
    //////////////////////////////////////////////////////////////////////////////////
    sut.update(<PickerContainer isVisible={false} />);
    const animatedView2 = sut.queryByTestId('containerAnimated');
    expect(sut).toMatchSnapshot('AnimatedView with invisible.');
    expect(animatedView2).toBeNull();
  });
});

describe('PickerContainer with all props', () => {
  it('should be applied properly', () => {
    const entering = ZoomIn.duration(500);
    const exiting = ZoomOut.duration(300);
    /**
     * entering/exitingを実行できなかったため、以下のPropsは検証できていません
     * - enteringCallback
     * - exitingCallback
     *
     * animatedPropsは取得できなかったため（Snapshot上にも存在していない）、検証できていません
     */
    const sut = render(
      <PickerContainer
        isVisible
        testID="animatedView"
        animatedProps={{pointerEvents: 'none'}}
        style={{backgroundColor: 'green'}}
        entering={entering}
        exiting={exiting}
      />,
    );
    expect(sut).toMatchSnapshot('PickerContainer with all props.');
    const animatedView = sut.getByTestId('animatedView');

    // assert animatedView
    const animatedViewProps = animatedView.props as Reanimated.AnimateProps<ViewProps>;
    expect(animatedViewProps.style).toEqual({backgroundColor: 'green'});
    expect(animatedViewProps.entering).toBe(entering);
    expect(animatedViewProps.exiting).toBe(exiting);
  });
});
