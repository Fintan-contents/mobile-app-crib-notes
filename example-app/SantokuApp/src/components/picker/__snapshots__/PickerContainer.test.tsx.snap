// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerContainer only with required props renders successfully only with required props: AnimatedView with invisible. 1`] = `null`;

exports[`PickerContainer only with required props renders successfully only with required props: AnimatedView with visible. 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {},
      }
    }
    collapsable={false}
    entering={
      SlideInDown {
        "build": [Function],
        "callbackV": [Function],
        "durationV": 300,
        "randomizeDelay": false,
      }
    }
    exiting={
      SlideOutDown {
        "build": [Function],
        "callbackV": [Function],
        "durationV": 300,
        "randomizeDelay": false,
      }
    }
    pointerEvents="box-none"
    style={Object {}}
    testID="containerAnimated"
  />
</View>
`;

exports[`PickerContainer with all props should be applied properly: PickerContainer with all props. 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-end",
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {},
      }
    }
    collapsable={false}
    entering={
      ZoomIn {
        "build": [Function],
        "callbackV": [Function],
        "durationV": 500,
        "randomizeDelay": false,
      }
    }
    exiting={
      ZoomOut {
        "build": [Function],
        "callbackV": [Function],
        "durationV": 300,
        "randomizeDelay": false,
      }
    }
    pointerEvents="box-none"
    style={
      Object {
        "backgroundColor": "green",
      }
    }
    testID="animatedView"
  />
</View>
`;
